### PR TITLE
Proposed Phase 2 API changes

### PR DESCRIPTION
**Pull request by jklontz**
The pull request contains the API changes proposed for phase 2. Here's the summary of changes:

- `janus_attributes_list` refactored as `janus_attributes`:
The new `janus_attributes` obviates the need for memory allocation on the implementation side, and thus obviates the need to introduce a function to deallocate the memory. It's also now easier for clients to retrieve the value of a specific attribute. The tradeoff is that the client must allocate memory to store all possible `JANUS_NUM_ATTRIBUTES` attributes, some of which may end up being unpopulated.

- `janus_detect` introduced:
This is now a prerequisite for `janus_augment`.

- `janus_track` dropped:
A change in the semantics of `janus_augment` obviated the need for this function.

- `JANUS_INVALID_ATTRIBUTE` enum dropped:
The change in semantics of `janus_attributes` obviated the need for this enum value.

- `JANUS_FRAME` enum dropped:
There isn't a clean way to propagate this information, and it doesn't seem to be needed anyway.

- `int` parameters in `janus_search` changed to `size_t`.
- Bumped version number to 0.4.0.
- Removed `iarpa_janus_aux.h` placeholder header file.

The evaluation harness will be updated after the API changes are accepted.

### CODE REVIEW

**Comment by jklontz** [comment]
@bhklein @sklum please review before I send this out to the mailing list. @bklare FYI.

File: include/iarpa_janus.h [hunk 1]
```diff
@@ -703,10 +720,10 @@ JANUS_EXPORT janus_error janus_search(const janus_flat_template probe,
                                       const size_t probe_bytes,
                                       const janus_flat_gallery gallery,
                                       const size_t gallery_bytes,
-                                      const int num_requested_returns,
+                                      const size_t num_requested_returns,
```
Comment (bhklein) [inline include/iarpa_janus.h]: Should this change be made in phase1 as well?

**Comment by bhklein** [comment]
looks good to me

File: include/iarpa_janus.h (hunk 1 above)
Comment (jklontz) [inline include/iarpa_janus.h]: It's mostly cosmetic, so I'm not inclined to back port it.

**Comment by sklum** [comment]
:+1:

**Comment by jklontz** [comment]
@biometrics/isi @biometrics/sri @biometrics/str @biometrics/umd Please review, thanks!

**Comment by bklare** [comment]
@jklontz One of the requirements we discussed was support for clustering. I
think this is a simple interface: gallery (input), num clusters (input,
optional), integer subject label assignment to each template (output).

Could you add this to this API?

**Comment by agrawalmoti** [comment]
I am guess we will have to solve the association problem for all the detects AND

Also associate one of the set of detections  to the person of interest using imagery (already enrolled images of that person in the gallery).

Correct me if I am wrong…

Moti

**Comment by jklontz** [comment]
@agrawalmoti Good question. As you've observed, the output of *janus_detect* is an array of *janus_attributes*, one for each detected face in the image, and the input to *janus_augment* is a single *janus_attributes*. In the event that there are multiple detections in an image, an ambiguity arises as to which one ought to be added to the template. For the purpose of evaluating algorithms, the evaluation harness will resolve this ambiguity by comparing the ground truth against the reported detection bounding boxes to select the appropriate detection. Having said that, the BAA does mention that we may _intentionally_ provide the incorrect face detection during template construction in order to measure your invariance to this source of error, so be aware of that possibility. In summary, we _won't_ ask you to solve the general detection association problem, but you should have some invariance to being given incorrect associations.

I did forget to add a gallery clustering function, will do so. My understanding with respect to the BAA is that the implementation of this function is considered _optional_.

**Comment by jklontz** [comment]
I've pushed a few more changes based on initial feedback:
- Added (optional) `janus_cluster` function for clustering.
- Further simplified `janus_attributes`, obviating the need for the `janus_attribute` enum.
- `janus_augment` is now allowed to modify the `attributes` parameter to return additional attribute information.
- Minor documentation corrections/improvements.

**Comment by jklontz** [comment]
Introduced `janus_attribute::frame_rate` to provide more information to `janus_augment`.

**Pull request merged**